### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by JoshDunnDev**
Implements issue #1.

### CODE REVIEW

File: WebAppManifestClient.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    } catch(err) {
+      if(!(manifestProxy && err.message.includes('CORS'))) {
+        throw err;
+      }
+    }
+    if(result) {
+      return result.data;
+    }
+    // proxy the request
+    // TODO: use different httpsAgent?
```
Comment (dlongley) [inline WebAppManifestClient.js]: Nope, use the same one.

File: WebAppManifestClient.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    manifestProxy = false
+  } = {}) {
+    this.defaultHeaders = {...DEFAULT_HEADERS, ...defaultHeaders};
+    this.httpsAgent = httpsAgent;
+    this.manifestProxyHost = manifestProxyHost;
+    this.manifestProxyPath = manifestProxyPath;
+    this.manifestProxy = manifestProxy;
+  }
+
+  async getManifest({origin}) {
```
Comment (dlongley) [inline WebAppManifestClient.js]: Can we have separate functions that get the manifest from the origin directly and that get it from the proxy as well -- so we can call those if desired? Then this function can call those helpers internally.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: WebAppManifestClient.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    manifestProxyHost,
+    manifestProxyPath = '/manifest.json'
+  } = {}) {
+    this.defaultHeaders = {...DEFAULT_HEADERS, ...defaultHeaders};
+    this.agent = agent;
+    this.manifestProxyHost = manifestProxyHost;
+    this.manifestProxyPath = manifestProxyPath;
+  }
+
+  async getManifest({origin}) {
```
Comment (dlongley) [inline WebAppManifestClient.js]: Here and on all public methods:

```suggestion
  async getManifest({origin} = {}) {
```

File: WebAppManifestClient.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        result = await this.getManifestFromProxy();
+        return result.data;
+      }
+      throw err;
+    }
+    return result.data;
+  }
+
+  async getManifestFromOrigin({origin}) {
+    const {defaultHeaders: headers, agent} = this;
```
Comment (dlongley) [inline WebAppManifestClient.js:45]:
```suggestion
    if(!(origin && typeof origin === 'string')) {
      throw new TypeError('"origin" must be a non-empty string.');
    }
    const {defaultHeaders: headers, agent} = this;
```

File: WebAppManifestClient.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  }) {
+    const {defaultHeaders: headers, agent} = this;
+    let manifest;
+    let icon = {};
+    try {
+      manifest = await this.getManifest({origin});
+    } catch(err) {
+      const favicon = await getFavicon({headers, agent, origin});
+      favicon ? icon.src = favicon : icon = defaultIcon;
+      return {manifest, icon};
```
Comment (dlongley) [inline WebAppManifestClient.js]: So, we shouldn't fetch the favicon, as that will require CORS to be used there as well. Rather, if a `defaultIcon` is given, we use that, if not, we use `icon.src = <faviconUrl>`. So the caller of the API will decide if they want a default icon or if they want it to default to using a favicon -- which will have to be detected as present later. Fetching a favicon in a Web UI happens via an `img` tag, which avoids CORs, so the model is a bit different from fetching the manifest.json file.

**Review by dlongley — CHANGES_REQUESTED** [review]
Looks like I never clicked submit on this review... submitting now.

File: WebAppManifestClient.js (hunk 5 above)
Comment (JoshDunnDev) [reply WebAppManifestClient.js]: So what if a favicon does not exist? Are you saying this would be detected at a different level than this?

Comment (dlongley) [reply WebAppManifestClient.js]: @JoshDunnDev, yes, a component that uses an `img` tag to show a favicon should have a fallback for a 404 case.  So -- this is handled at a higher level.

**Comment by dlongley** [comment]
@JoshDunnDev -- What's the status here? It looks like a review was never re-requested. Did this work get finished and is it ready for review?

File: WebAppManifestClient.js [hunk 6]
```diff
@@ -70,39 +73,32 @@ export class WebAppManifestClient {
...
       manifest = await this.getManifest({origin});
     } catch(err) {
-      const favicon = await getFavicon({headers, agent, origin});
-      favicon ? icon.src = favicon : icon.src = defaultIcon;
-      if(!icon.src) {
-        icon = undefined;
-      }
+      defaultIcon ? icon.src = defaultIcon : icon.src =
+        `${origin}/favicon.ico`;
       return {manifest, icon};
```
Comment (dlongley) [inline WebAppManifestClient.js]: This is simpler, right?

```suggestion
      icon.src = defaultIcon || `${origin}/favicon.ico`;
      return {manifest, icon};
```

File: WebAppManifestClient.js [hunk 7]
```diff
@@ -70,39 +73,32 @@ export class WebAppManifestClient {
...
+    colorScheme = 'light', defaultIcon, manifest, origin, size
+  } = {}) {
+    let icon = {};
     icon = await getWebAppManifestIcon({colorScheme, manifest, origin, size});
     if(!icon) {
       icon = {};
-      const favicon = await getFavicon({headers, agent, origin});
-      favicon ? icon.src = favicon : icon.src = defaultIcon;
+      defaultIcon ? icon.src = defaultIcon : icon.src =
+        `${origin}/favicon.ico`;
```
Comment (dlongley) [inline WebAppManifestClient.js]:
```suggestion
      icon.src = defaultIcon || `${origin}/favicon.ico`;
```

**Review by dlongley — CHANGES_REQUESTED** [review]

**Pull request merged**